File: 백준/Silver/15652. N과 M （4）/N과 M （4）.js
const readline = require('readline');
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

let N, M;
const rs = [];

function recur(num, start) {
    if (num === M) {
        console.log(rs.join(' '));
        return;
    }
    
    for (let i = start; i <= N; i++) {
        rs.push(i);
        recur(num + 1, i);
        rs.pop();
    }
}

rl.on('line', function(line) {
    [N, M] = line.split(' ').map(Number);
    recur(0, 1);
    rl.close();
});

File: 백준/Silver/15652. N과 M （4）/N과 M （4）.py
# 15652 N과 M (4)
# 자연수 N과 M이 주어졌을 때, 
# 아래 조건을 만족하는 길이가 M인 수열을 모두 구하는 프로그램을 작성하시오.
N, M = map(int, input().split())
rs = []

def recur(num, start):  # start 매개변수 추가
    if num == M:
        print(' '.join(map(str, rs)))
        return
    
    for i in range(start, N+1):  # start부터 시작
        rs.append(i)
        recur(num+1, i)  # 다음 숫자는 현재 숫자(i)부터 시작
        rs.pop()

recur(0, 1)